File: LeetCode/Hard/2014-Longest_Subsequence_Repeated_K_Times/2014-LongestSubsequenceRepeatedKTimes.js
var is_subsequence = function (s, subseq, k) {
    let i = 0;
    let j = 0;

    while (i < s.length) {
        if (s[i] == subseq[j]) {
            j++;

            if (j == subseq.length) {
                j = 0;
                k--;

                if (k == 0)
                    return true;
            }
        }
        i++;
    }

    return false;
}


/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var longestSubsequenceRepeatedK = function (s, k) {
    const freq = new Map();

    for (const char of s) {
        freq.set(char, (freq.get(char) || 0) + 1);
    }

    const possible_chars = [];

    for (const [key, value] of freq) {
        if (value >= k)
            possible_chars.push(key)
    }

    possible_chars.sort();

    let queue = [""];
    let ans;

    while (queue.length != 0) {
        let curr_subseq = queue.shift();
        ans = curr_subseq;

        for (let c of possible_chars) {
            let candidate = curr_subseq + c;
            if (is_subsequence(s, candidate, k))
                queue.push(candidate);
        }
    }

    return ans
};

File: LeetCode/Hard/2014-Longest_Subsequence_Repeated_K_Times/2014-LongestSubsequenceRepeatedKTimes.ex
defmodule Solution do
  @spec longest_subsequence_repeated_k(s :: String.t, k :: integer) :: String.t
  def longest_subsequence_repeated_k(s, k) do
    freq = s
      |> String.graphemes()
      |> Enum.reduce(%{}, fn char, acc ->
      Map.update(acc, char, 1, fn count -> count + 1 end)
    end)

    possible_chars = Enum.to_list(freq)
      |> Enum.filter(fn {_, count} -> count >= k end)
      |> Enum.map(fn {char, _} -> char end)
      |> Enum.sort()

    queue = [""]
    ans = ""

    bfs(queue, ans, s, k, possible_chars)
  end

  defp bfs([], ans, _s, _k, _possible_chars) , do: ans

  defp bfs([curr_subseq | rest], ans, s, k, possible_chars) do
    new_ans = curr_subseq

    next =
      for c <- possible_chars do
        candidate = curr_subseq <> c
        if is_subsequence(s, candidate, k) do
          candidate
        else
          nil
        end
      end
      |> Enum.filter(&(&1 != nil))

    new_queue = rest ++ next

    bfs(new_queue, new_ans, s, k, possible_chars)
  end

  defp is_subsequence(s, subseq, k) do
    target = String.duplicate(subseq, k)

    check(String.to_charlist(s), String.to_charlist(target))
  end

  # This is the standard two-pointer/recursive subsequence check for charlists.
  defp check(_s, []) do
    true # All target characters found
  end
  defp check([], _target) do
    false # Target characters remaining but no more source characters
  end
  defp check([s_head | s_tail], [t_head | t_tail]) do
    if s_head == t_head do
      # Match: advance both
      check(s_tail, t_tail)
    else
      # No match: advance only source
      check(s_tail, [t_head | t_tail])
    end
  end
end
